### PR TITLE
Closes #1218 Extend `pdarray` setops to work for multiple arrrays

### LINKED ISSUE #1218: Extend `pdarray` setops to work for multiple arrays
Write a PR that extends the ordinary set ops to work on multiple arrays, much like alignment.py:in1d_multi(a, b) is the multi-array analog of pdarraysetops.py:in1d(a, b). Stylistically, I think the logic of in1d_multi should be rolled into in1d, which should accept either singleton arrays or tuples/lists of arrays as arguments. Similar logic should suffice to make the other set operations work for multiple arrays.

Adds support for multiple arrays to the existing `pdarray` set operations:
- `union1d`
- `intersection1d`
- `setdiff1d`
- `setxor1d`

As an example, use `alignment.py:in1d_multi(a, b)` as a guide. It is the multi-array analog of `pdarraysetops.py:in1d(a, b)`.

@reuster986 mentioned a preference for the logic to be consolidated into a single function that takes in single arrays or tuples/lists of arrays as arguments.

Example utilization:

```python
def set_intx(cls, a, b):
    a_seg_inds = a.grouping.broadcast(arange(a.size))
    b_seg_inds = b.grouping.broadcast(arange(b.size))
    (new_seg_inds, new_values) = intersect1d((a_seg_inds, a.values), (b_seg_inds, b.values))
    g = ak.GroupBy(new_seg_inds)
    # TODO: Possibly add logic to insert empty segments where intersection is empty
    return SegArray(g.segments, new_values[g.permutation])
```

This issue is needed to make the work for #1197 (PR #1213) better performing. That PR has been converted to draft until this work is completed. Once completed, the PR will need to be modified to utilize the updated `pdarray` set functionality.

### PR DESCRIPTION
**Pull request by Ethan-DeBandi99**
This PR (closes #1218):

Adds support for multiple arrays to the existing `pdarray` set operations:
- `union1d`
- `intersect1d`
- `setdiff1d`
- `setxor1d`
*Note - `in1d` multiple array support is defined in `arkouda/alignment.py:in1d_multi(a, b)`. This will need to be moved separately.*

Adds support for set operations to SegArrays:
- `union_ma`
- `intersect_ma`
- `setdiff_ma`
- `setxor_ma`
*Note - the `_ma` suffix is denoting that these are utilizing the multiple array code added to `pdarraysetops.py`. This was added to prevent conflicts if PR #1213 was merged before this PR was submitted. I am suggesting that PR be closed as this is a better solution. If everyone agrees, I will update the naming to remove the `_ma` suffix.*

Only int64 and uint64 are currently supported for these set operations.

Testing for:
- Multiple Array `pdarray` Setops - `tests/setops_test.py` (added `_multi` functions)
- `SegArray` Setops - `tests/segarray_test.py` (added setops tests)

Benchmarks for:
- Multiple Array `pdarray` setops - `benchmarks/setops_multiarray.py`
- `SegArray` Setops - `benchmarks/segarray_setops.py`

The current implementation is all client side. I discussed with @reuster986 and we plan to add an issue to move this to the server in the future.

`pdarray` set operations have been updated to take a single `pdarray` for each element or a `list` of `pdarrays` for each argument. The later triggers the multiple array workflow. This requires  the lists be the same size and for the elements within a list to be equal size.

```python
# Example Multiple Array Setops 
a = [1, 2, 3, 4, 5]
a1 = ak.array(a)
b1 = ak.array([1, 5, 2, 3, 4])
b2 = ak.array([1, 3, 2, 5, 4])

intx = ak.intersect1d([a1, a1], [b1, b2])
intx
[array([1 2 4]), array([1 2 4])]

# Example SegArray Setop
a = [1, 2, 3, 4, 5]
b = [6, 7, 8]
c = [1, 2, 4]
d = [8]
segarr = ak.SegArray(ak.array([0, len(a)]), ak.array(a + b))
segarr_2 = ak.SegArray(ak.array([0, len(c)]), ak.array(c + d))

intx = segarr.intersect_ma(segarr_2)
SegArray([
[1 2 4]
[8]
])
```

### CODE REVIEW

File: arkouda/pdarraysetops.py [hunk 1]
```diff
@@ -436,26 +516,64 @@ def setdiff1d(pda1: pdarray, pda2: pdarray,
...
+        g = GroupBy(c)
+        if assume_unique:
+            # need to verify uniqueness, otherwise answer will be wrong
+            if (g.sum(isa)[1] > 1).any():
+                raise ValueError("Called with assume_unique=True, but first argument is not unique")
+            if (g.sum(~isa)[1] > 1).any():
+                raise ValueError("Called with assume_unique=True, but second argument is not unique")
+        k, ct = g.count()
+        truth = g.broadcast(ct == 1, permute=True)
+        return [x[truth[isa]] for x in ua]
```
Comment (reuster986) [inline arkouda/pdarraysetops.py:568]: Similar to above, storing `truth[isa]` in a variable might save time and avoid recomputation inside the loop (again, not totally sure -- python might be smart enough to cache).

File: arkouda/pdarraysetops.py [hunk 2]
```diff
@@ -495,21 +613,61 @@ def setxor1d(pda1: pdarray, pda2: pdarray,
...
+            return create_pdarray(cast(str, repMsg))
+        if not assume_unique:
+            pda1 = cast(pdarray, unique(pda1))
+            pda2 = cast(pdarray, unique(pda2))
+        aux = concatenate((pda1, pda2), ordered=False)
+        aux_sort_indices = argsort(aux)
+        aux = aux[aux_sort_indices]
+        flag = concatenate((array([True]), aux[1:] != aux[:-1], array([True])))
+        return aux[flag[1:] & flag[:-1]]
+    elif isinstance(pda1, list) and isinstance(pda2, list):
```
Comment (reuster986) [inline arkouda/pdarraysetops.py:645]: tuples too :)

File: arkouda/pdarraysetops.py [hunk 3]
```diff
@@ -495,21 +613,61 @@ def setxor1d(pda1: pdarray, pda2: pdarray,
...
+        c = [concatenate(x, ordered=False) for x in zip(ua, ub)]
+        g = GroupBy(c)
+        if assume_unique:
+            # need to verify uniqueness, otherwise answer will be wrong
+            if (g.sum(isa)[1] > 1).any():
+                raise ValueError("Called with assume_unique=True, but first argument is not unique")
+            if (g.sum(~isa)[1] > 1).any():
+                raise ValueError("Called with assume_unique=True, but second argument is not unique")
+        k, ct = g.count()
+        truth = g.broadcast(ct == 1, permute=True)
```
Comment (reuster986) [inline arkouda/pdarraysetops.py:668]: At this point, instead of broadcasting and re-grouping, you can just do:

```python
single = ct == 1
return [x[single] for x in k]
```

File: arkouda/segarray.py [hunk 4]
```diff
@@ -803,5 +803,213 @@ def load(cls, prefix_path, dataset='segarray', segment_suffix='_segments', value
...
+        # This method does not return any empty resulting segments. We need to add these if they are missing.
+        if g.segments.size == self.size:
+            return SegArray(g.segments, new_values[g.permutation])
+        else:
+            segments = zeros(self.size, dtype=akint64)
+            k, ct = g.count()
+            segments[k] = g.segments
+            truth = ~self._non_empty | ~other._non_empty
+            if truth[-1]:
+                segments[-1] = new_values[g.permutation].size
```
Comment (reuster986) [inline arkouda/segarray.py:853]: Since you only need the size, you could eliminate the gather by doing `segments[-1] = g.permutation.size`

File: arkouda/segarray.py [hunk 5]
```diff
@@ -803,5 +803,213 @@ def load(cls, prefix_path, dataset='segarray', segment_suffix='_segments', value
...
```
Comment (reuster986) [inline arkouda/segarray.py:905]: See above

File: arkouda/segarray.py [hunk 6]
```diff
@@ -803,5 +803,213 @@ def load(cls, prefix_path, dataset='segarray', segment_suffix='_segments', value
...
```
Comment (reuster986) [inline arkouda/segarray.py:957]: See above

File: arkouda/segarray.py [hunk 7]
```diff
@@ -803,5 +803,213 @@ def load(cls, prefix_path, dataset='segarray', segment_suffix='_segments', value
...
```
Comment (reuster986) [inline arkouda/segarray.py:1009]: See above

**Review by reuster986 — CHANGES_REQUESTED** [review]
The logic looks great! And thank you for handling empty segments and creating the benchmarks! Just a few minor requests in the comments.

File: arkouda/pdarraysetops.py [hunk 8]
```diff
@@ -311,45 +328,71 @@ def union1d(pda1: pdarray, pda2: pdarray) -> pdarray:
...
+            return pda1  # union is pda1
+        if pda1.dtype == int and pda2.dtype == int or \
+                (pda1.dtype == akuint64 and pda2.dtype == akuint64):
+            repMsg = generic_msg(cmd="union1d", args="{} {}".
+                                 format(pda1.name, pda2.name))
+            return cast(pdarray, create_pdarray(repMsg))
+        return cast(pdarray,
+                    unique(cast(pdarray,
+                                concatenate((unique(pda1), unique(pda2)), ordered=False))))  # type: ignore
+    elif isinstance(pda1, list) and isinstance(pda2, list):
```
[the remark this answers is not in the document]
Comment (Ethan-DeBandi99) [reply arkouda/pdarraysetops.py:358]: Updated the functions to allow parameters to be Tuples. Updated the second if to require both elements to be either list or tuple. The processing for each should be the same. I also added test cases to verify that list and tuples are functional for all operations.

File: arkouda/pdarraysetops.py [hunk 9]
```diff
@@ -368,52 +411,89 @@ def intersect1d(pda1: pdarray, pda2: pdarray,
...
+        if not assume_unique:
+            pda1 = unique(pda1)
+            pda2 = unique(pda2)
+        aux = concatenate((pda1, pda2), ordered=False)
+        aux_sort_indices = argsort(aux)
+        aux = aux[aux_sort_indices]
+        mask = aux[1:] == aux[:-1]
+        int1d = aux[:-1][mask]
+        return int1d
+    elif isinstance(pda1, list) and isinstance(pda2, list):
```
[the remark this answers is not in the document]
Comment (Ethan-DeBandi99) [reply arkouda/pdarraysetops.py:447]: See Comment above.

File: arkouda/pdarraysetops.py [hunk 10]
```diff
@@ -368,52 +411,89 @@ def intersect1d(pda1: pdarray, pda2: pdarray,
...
+        c = [concatenate(x, ordered=False) for x in zip(ua, ub)]
+        g = GroupBy(c)
+        if assume_unique:
+            # need to verify uniqueness, otherwise answer will be wrong
+            if (g.sum(isa)[1] > 1).any():
+                raise ValueError("Called with assume_unique=True, but first argument is not unique")
+            if (g.sum(~isa)[1] > 1).any():
+                raise ValueError("Called with assume_unique=True, but second argument is not unique")
+        k, ct = g.count()
+        return [x[ct == 2] for x in k]
```
[the remark this answers is not in the document]
Comment (Ethan-DeBandi99) [reply arkouda/pdarraysetops.py:470]: I am not sure it python caches this either. To be sure, I add the variable as suggested and utilize that in the loop.

File: arkouda/pdarraysetops.py [hunk 11]
```diff
@@ -436,26 +516,64 @@ def setdiff1d(pda1: pdarray, pda2: pdarray,
...
+        if (pda1.dtype == int and pda2.dtype == int) or \
+                (pda1.dtype == akuint64 and pda2.dtype == akuint64):
+            repMsg = generic_msg(cmd="setdiff1d", args="{} {} {}". \
+                                 format(pda1.name, pda2.name, assume_unique))
+            return create_pdarray(cast(str, repMsg))
+        if not assume_unique:
+            pda1 = cast(pdarray, unique(pda1))
+            pda2 = cast(pdarray, unique(pda2))
+        return pda1[in1d(pda1, pda2, invert=True)]
+    elif isinstance(pda1, list) and isinstance(pda2, list):
```
[the remark this answers is not in the document]
Comment (Ethan-DeBandi99) [reply arkouda/pdarraysetops.py:544]: See previous comment.

File: arkouda/pdarraysetops.py (hunk 1 above)
Comment (Ethan-DeBandi99) [reply arkouda/pdarraysetops.py:568]: To be safe, I updated the code to store `truth[isa]` to a var and call that in the loop.

File: arkouda/pdarraysetops.py (hunk 3 above)
Comment (Ethan-DeBandi99) [reply arkouda/pdarraysetops.py:668]: Updated to use your suggestion. This was the one I was hoping you'd have a suggestion for. Thanks!

Comment (Ethan-DeBandi99) [reply arkouda/pdarraysetops.py:668]: Updated to eliminate the gather.

File: arkouda/segarray.py (hunk 4 above)
Comment (Ethan-DeBandi99) [reply arkouda/segarray.py:853]: Updated to eliminate the gather.

File: arkouda/segarray.py (hunk 5 above)
Comment (Ethan-DeBandi99) [reply arkouda/segarray.py:905]: Updated to eliminate the gather.

File: arkouda/segarray.py (hunk 6 above)
Comment (Ethan-DeBandi99) [reply arkouda/segarray.py:957]: Updated to eliminate the gather.

File: arkouda/segarray.py (hunk 7 above)
Comment (Ethan-DeBandi99) [reply arkouda/segarray.py:1009]: Updated to eliminate the gather.

File: arkouda/segarray.py [hunk 12]
```diff
@@ -803,5 +803,213 @@ def load(cls, prefix_path, dataset='segarray', segment_suffix='_segments', value
...
+        else:
+            segments = zeros(self.size, dtype=akint64)
+            k, ct = g.count()
+            segments[k] = g.segments
+            truth = ~self._non_empty | ~other._non_empty
+            if truth[-1]:
+                segments[-1] = g.permutation.size
+                truth[-1] = False
+            segments[truth] = segments[arange(self.size)[truth] + 1]
+            return SegArray(segments, new_values[g.permutation])
```
Comment (pierce314159) [inline arkouda/segarray.py]: found a small case for `setdiff` with multiple empty segments that isn't correctly handled. I let ethan know and he has a fix

File: benchmarks/segarray_setops.py [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+            r = fxn(seg_b)
+            end = time.time()
+            timings[op].append(end - start)
+            results[op] = r
+    tavg = {op: sum(t) / trials for op, t in timings.items()}
+
+    for op, t in tavg.items():
+        print("  {} Average time = {:.4f} sec".format(op, t))
+        # bytes_per_sec = (a.size * a.itemsize * 2) / t
+        # print("  {} Average rate = {:.2f} GiB/sec".format(op, bytes_per_sec / 2 ** 30))
```
Comment (pierce314159) [inline benchmarks/segarray_setops.py:47]: was the average rate code meant to be left commented?

File: benchmarks/setops_multiarray.py [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+        fxn = getattr(npa, op_map[op])
+        npr = list(fxn(npb))
+        if len(npr) > 0:
+            npr0, npr1 = map(list, zip(*npr))
+        else:
+            npr0 = []
+            npr1 = []
+        fxn = getattr(ak, op)
+        akr = fxn(pd_list_a, pd_list_b)
+        print(f"\nNPR0: {npr0}\n\nNPR1: {npr1}\n\nAKR: {akr}\n")
```
Comment (pierce314159) [inline benchmarks/setops_multiarray.py:75]: was this print statement left in intentionally?

File: benchmarks/setops_multiarray.py [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+        for dtype in TYPES:
+            check_correctness(dtype, args.seed)
+        sys.exit(0)
+
+    print("array size = {:,}".format(args.size))
+    print("number of trials = ", args.trials)
+    time_ak_setops(args.size, args.trials, args.dtype, args.seed)
+    if args.numpy:
+        time_np_setops(args.size, args.trials, args.dtype, args.seed)
+        print("Verifying agreement between arkouda and NumPy on small problem... ", end="")
```
Comment (pierce314159) [inline benchmarks/setops_multiarray.py:117]: there isn't a `time_np_setops` in the `setops_multiarray` benchmark. The `--numpy` flag also needs to be removed from the parser

**Review by pierce314159 — COMMENTED** [review]
Everything looks good! The logic is really neat

Some very minor comments mostly about docstring examples and things left in/out of the benchmarks

I agree with updating the names to remove the `_ma`

File: arkouda/pdarraysetops.py [hunk 16]
```diff
@@ -311,45 +330,72 @@ def union1d(pda1: pdarray, pda2: pdarray) -> pdarray:
...
+            return pda1  # union is pda1
+        if pda1.dtype == int and pda2.dtype == int or \
+                (pda1.dtype == akuint64 and pda2.dtype == akuint64):
+            repMsg = generic_msg(cmd="union1d", args="{} {}".
+                                 format(pda1.name, pda2.name))
+            return cast(pdarray, create_pdarray(repMsg))
+        return cast(pdarray,
+                    unique(cast(pdarray,
+                                concatenate((unique(pda1), unique(pda2)), ordered=False))))  # type: ignore
+    elif (isinstance(pda1, list) and isinstance(pda2, list)) or (isinstance(pda1, tuple) and isinstance(pda2, tuple)):
```
[the remark this answers is not in the document]
Comment (Ethan-DeBandi99) [reply arkouda/pdarraysetops.py]: I updated the if block. In my head it made sense for both to be the same type, but since both are iterable, this should be allowed.

File: arkouda/pdarraysetops.py [hunk 17]
```diff
@@ -311,45 +330,72 @@ def union1d(pda1: pdarray, pda2: pdarray) -> pdarray:
...
+
+    #Multi-Array Example
+    >>> a = ak.arange(5)
+    >>> b = ak.array([1, 5, 3, 4, 2])
+    >>> c = ak.array([1, 4, 3, 2, 5])
+    >>> d = ak.array([1, 2, 3, 5, 4])
+    >>> multia = [a, a, a]
+    >>> multib = [b, c, d]
+    >>> ak.union1d(multia, multib)
+    [array[1, 2, 2, 3, 4, 4, 5, 5], array[1, 2, 5, 3, 2, 4, 4, 5], array[1, 2, 4, 3, 4, 5, 2, 5]]
```
[the remark this answers is not in the document]
Comment (Ethan-DeBandi99) [reply arkouda/pdarraysetops.py]: updated. I copied the setup code from the wrong location. Thanks for catching.

File: arkouda/pdarraysetops.py [hunk 18]
```diff
@@ -311,45 +330,72 @@ def union1d(pda1: pdarray, pda2: pdarray) -> pdarray:
...
+        ua = ag.unique_keys
+        bg = GroupBy(pda2)
+        ub = bg.unique_keys
+
+        c = [concatenate(x, ordered=False) for x in zip(ua, ub)]
+        g = GroupBy(c)
+        k, ct = g.count()
+        return k
+    else:
+        raise TypeError('Both pda1 and pda2 must be pdarray or list')
```
[the remark this answers is not in the document]
Comment (Ethan-DeBandi99) [reply arkouda/pdarraysetops.py]: Updated to include tuple and print the types that were supplied.

File: arkouda/pdarraysetops.py [hunk 19]
```diff
@@ -436,26 +521,66 @@ def setdiff1d(pda1: pdarray, pda2: pdarray,
     >>> b = ak.array([3, 4, 5, 6])
     >>> ak.setdiff1d(a, b)
     array([1, 2])
+
+    #Multi-Array Example
+    >>> a = ak.arange(5)
```
[the remark this answers is not in the document]
Comment (Ethan-DeBandi99) [reply arkouda/pdarraysetops.py]: Updated

File: arkouda/pdarraysetops.py [hunk 20]
```diff
@@ -495,21 +620,59 @@ def setxor1d(pda1: pdarray, pda2: pdarray,
...
     array([1, 4, 5, 7])
+
+    #Multi-Array Example
+    >>> a = ak.arange(5)
+    >>> b = ak.array([1, 5, 3, 4, 2])
+    >>> c = ak.array([1, 4, 3, 2, 5])
+    >>> d = ak.array([1, 2, 3, 5, 4])
+    >>> multia = [a, a, a]
+    >>> multib = [b, c, d]
+    >>> ak.setdiff1d(multia, multib)
```
[the remark this answers is not in the document]
Comment (Ethan-DeBandi99) [reply arkouda/pdarraysetops.py]: Updated

File: benchmarks/segarray_setops.py [hunk 21]
```diff
@@ -0,0 +1,111 @@
...
+        seg_b = ak.SegArray(ak.array([0, len(c)]), ak.concatenate([c, d]))
+    elif dtype == 'uint64':
+        a = ak.randint(0, 2 ** 32, N, seed=seed, dtype=ak.uint64)
+        b = ak.randint(0, 2 ** 32, N, seed=seed, dtype=ak.uint64)
+        seg_a = ak.SegArray(ak.array([0, len(a)]), ak.concatenate[[a, b]])
+        c = ak.randint(0, 2 ** 32, N, seed=seed, dtype=ak.uint64)
+        d = ak.randint(0, 2 ** 32, N, seed=seed, dtype=ak.uint64)
+        seg_b = ak.SegArray(ak.array([0, len(c)]), ak.concatenate([c, d]))
+
+    print(OPS)
```
[the remark this answers is not in the document]
Comment (Ethan-DeBandi99) [reply benchmarks/segarray_setops.py]: No it was for testing, it has been removed.

File: benchmarks/segarray_setops.py (hunk 13 above)
Comment (Ethan-DeBandi99) [reply benchmarks/segarray_setops.py]: Nope, uncommented it

File: benchmarks/setops_multiarray.py (hunk 14 above)
Comment (Ethan-DeBandi99) [reply benchmarks/setops_multiarray.py]: No I did not cleanup everything from my testing. Removed the print

File: benchmarks/setops_multiarray.py (hunk 15 above)
Comment (Ethan-DeBandi99) [reply benchmarks/setops_multiarray.py:114]: Removed the flag.

File: arkouda/pdarraysetops.py [hunk 22]
```diff
@@ -275,7 +275,8 @@ def concatenate(arrays: Sequence[Union[pdarray, Strings, 'Categorical', ]],  # t
         raise TypeError('arrays must be an array of pdarray or Strings objects')
 
 
-def multiarray_setop_validation(pda1: List[pdarray], pda2: List[pdarray]):
+def multiarray_setop_validation(pda1: Union[List[pdarray], Tuple[pdarray, ...]],
```
Comment (reuster986) [inline arkouda/pdarraysetops.py:278]: Sorry for just noticing this, but should we make these `Union[List[groupable], Tuple[groupable]]`? I'm 95% sure all the logic will work with Strings as well as pdarrays, but maybe we should leave that for a separate PR, because it would require additional tests.

**Review by reuster986 — APPROVED** [review]

File: arkouda/pdarraysetops.py (hunk 22 above)
Comment (Ethan-DeBandi99) [reply arkouda/pdarraysetops.py:278]: No problem. I will add an Issue to update the typing and add additional testing.

Comment (Ethan-DeBandi99) [reply arkouda/pdarraysetops.py:278]: Issue #1280 has been added to track this.

**Review by pierce314159 — APPROVED** [review]
